server/aesdsocket.c

```diff
@@ -0,0 +1,372 @@
+            else
+                server_socket_state = STATE_LISTENING;
+            break;
+
+        case STATE_START_DAEMON:
+            // LOG_DBG("DAEMON\n");
+            signal(SIGCHLD, SIG_IGN);
+            signal(SIGHUP, SIG_IGN);
+            server_socket_state = STATE_LISTENING;
+
+            pid = fork();
+            if (pid == -1)
+                return -1;
+            else if (pid != 0)
+                exit(EXIT_SUCCESS);
+
+            umask(0);
+
+            if (setsid() == -1)
+            {
+                perror("Error starting daemon");
+                return_status = -1;
+                server_socket_state = EXIT;
+                break;
+            }
+
+            if (chdir("/") == -1)
+            {
+                perror("Error starting daemon");
+                return_status = -1;
+                server_socket_state = EXIT;
+                break;
+            }
+
+            close(STDIN_FILENO);
+            close(STDOUT_FILENO);
+            close(STDERR_FILENO);
+
+            //redirect files to /dev/null
+            open("/dev/null", O_RDWR);
+            dup(0);
+            dup(0);
+
+            break;
+
+        case STATE_LISTENING:
+            // LOG_DBG("STATE_LISTENING\n");
+            //listen for connections
+            status = listen(sockfd_server, 5);
+            if (status < 0)
+            {
+                perror("Listening failed");
+                return_status = -1;
+                server_socket_state = EXIT;
+                break;
+            }
+            server_socket_state = STATE_OPENING_FILE;
+            break;
+
+        case STATE_OPENING_FILE:
+        //    LOG_DBG("STATE_OPENING_FILE\n");
+           fd = open("/var/tmp/aesdsocketdata", O_CREAT | O_RDWR | O_TRUNC, 0644);
+            if (fd == -1)
+            {
+                perror("error creating file");
+                return_status = -1;
+                server_socket_state = EXIT;
+                break;
+            }
+            server_socket_state = STATE_ACCEPTING;
+            break;
+
+        case STATE_ACCEPTING:
+            // LOG_DBG("STATE_ACCEPTING\n");
+            addr_size = sizeof(new_addr);
+		
+            //accept socket connections
+            sockfd_client = accept(sockfd_server, (struct sockaddr*)&new_addr, &addr_size);
+            if(sockfd_client < 0) 
+            {
+                LOG_ERROR("error accepting connection");
+                server_socket_state = EXIT;
+                break;
+            }
+
+            inet_ntop(AF_INET, &client_addr, str, INET_ADDRSTRLEN);
+            
+		    syslog(LOG_INFO, "Accepted connection from %s", str);
+
+
+            nwrite=0, nreceived=0, nsent=0, nread=0;
+            heap_buffer_for_write = NULL;
+            heap_buffer_for_read = NULL;
+            filled_up_buffer_size = 0;
+            total_buffer_size = BUF_SIZE;
+
+            heap_buffer_for_write = (char *)malloc(sizeof(char)*BUF_SIZE);
```

Comment:
@rajatchaple you wouldn't need to dynamically allocate this buffer since you always use a fixed size.